pkg/apis/monitoring/v1/alertmanager_types.go

```diff
@@ -270,6 +270,8 @@ type AlertmanagerConfiguration struct {
 // AlertmanagerGlobalConfig configures parameters that are valid in all other configuration contexts.
 // See https://prometheus.io/docs/alerting/latest/configuration/#configuration-file
 type AlertmanagerGlobalConfig struct {
+	*GlobalSMTPConfig `json:",inline"`
```

Comment:
I'd prefer to avoid type embedding and suggest to add all SMTP fields directly to the `AlertmanagerGlobalConfig` struct.

Reply:
My reasons for using embeds:
1. The embedding method gathers a group of related configurations, which is convenient for writing `convertSMTPConfig`, and is convenient for being called in the `convertGlobalConfig` function, avoiding the function body being too long and making it more readable
2. In addition to automatically generating more codes, the embedded structure seems to have no other disadvantages. If so, I am sorry（QAQ

If you insist, then I would avoid embedding

Reply:
Embedding can create subtle hard-to-debug issues (though in practice using it here is probably safe in the long term).

Another option would be to introduce a specific field grouping all SMTP parameters together: 

```
type AlertmanagerGlobalConfig struct {
	SMTPConfig *GlobalSMTPConfig `json:"smtp,omitempty"`
...
}

type GlobalSMTPConfig struct {
	From *string `json:"from,omitempty"`
...
}
```

The manifest would look like:

```
spec:
  alertmanagerConfiguration:
    name: foo
    global:
      smtp:
        from: bar@example.com
        requireTLS: true
```

Reply:
Thanks for letting me know this.

This haved been discussed at [4606#issuecomment-1058087736](https://github.com/prometheus-operator/prometheus-operator/issues/4606#issuecomment-1058087736).

I prefer a hierarchical structure, but it's different from the flat structure of the Alertmanager configuration. While I think it's acceptable and clear, users need to read [prometheus-operator's AlertmanagerGlobalConfig documentation](https://prometheus-operator.dev/docs/operator/api/#monitoring.coreos.com/v1.AlertmanagerGlobalConfig) before configuring.

I expect you to have a strong opinion this time 😂, both fine by me. You decide, I'll do it

Reply:
:rofl: 

As of today, I'm less convinced that being close to the Alertmanager configuration format is the right choice. In any case, the field names aren't the same (e.g. `smtp_from` vs. `smtpFrom`) so you're not going to be able to copy/paste your original Alertmanager config file into the custom resource. For this reason, I'd go the way suggested in https://github.com/prometheus-operator/prometheus-operator/pull/5649#discussion_r1224432215.